File: src/multi/model/SubProblemReactive.java
package multi.model;

import ilog.concert.*;
import ilog.cplex.IloCplex;
import multi.data.InputData;
import multi.data.IntArrayWrapper;
import multi.data.Parameter;
import multi.structure.Request;

public class SubProblemReactive extends BasePrimalModel
{
    private int [][] vVarValue1;
    private int [][] vVarValue2;
    private double [] uu;
    private double ladenCost;
    private double emptyCost;
    private double penaltyCost;
    private double rentalCost;

    public SubProblemReactive(InputData in, Parameter p){
        super();
        this.in = in;
        this.p = p;
        this.ModelName = "SPR"+ "-R"+ in.getShipRouteSet().size() + "-T" + p.getTimeHorizon() + "-"+ FleetType + "-S" + randomSeed;
        this.vVarValue1= new int[p.getVesselSet().length] [p.getShippingRouteSet().length];
        this.vVarValue2= new int[p.getVesselSet().length] [p.getVesselPathSet().length];
        this.uValue = new double[p.getDemand().length];
        try{
            cplex = new IloCplex();
            publicSetting(cplex);
            frame();
        }catch (IloException e) {
            e.printStackTrace();
        }
    }

    protected void setDecisionVars() throws IloException{
        SetRequestDecisionVars();
    }

    // Minimize total cost about containers
    protected void setObjectives() throws IloException	{
        IloLinearNumExpr Obj=cplex.linearNumExpr();

        // i
        for(int i=0;i<p.getDemand().length;i++)
        {
            // item2 : Penalty Cost of unsatisfied Demand : c2i*Gi
            Obj.addTerm(p.getPenaltyCostForDemand()[i], gVar[i]);

            Request od = in.getRequestSet().get(i);
            // \phi_i
            for(int k=0;k<od.getNumberOfLadenPath();k++)
            {
                int j = od.getLadenPathIndexes()[k];
                // item3 : Demurrage of self-owned and leased containers and Rental cost on laden paths
                Obj.addTerm(p.getLadenPathCost()[j], xVar.get(i)[k]);
                Obj.addTerm(p.getLadenPathCost()[j], yVar.get(i)[k]);
                Obj.addTerm(p.getRentalCost()*p.getTravelTimeOnPath()[j], yVar.get(i)[k]);
            }

            // item4: Demurrage of self-owned containers for empty path
            // \theta
            for(int k=0;k<od.getNumberOfEmptyPath();k++)
            {
                int j = od.getEmptyPathIndexes()[k];

                Obj.addTerm(p.getEmptyPathCost()[j], zVar.get(i)[k]);
            }
        }

        cplex.addMinimize(Obj);
    }

    private IloRange[] C1;
    private IloRange[] C2_1;
    private IloRange[] C2_2;
    private IloRange[][] C3;

    protected void setConstraints() throws IloException {
        setConstraint1();
        setConstraint2_1();
        setConstraint2_2();
        setConstraint3();
    }

    // Demand equation :
    // C-5------α
    // C-5 = 0
    private void setConstraint1() throws IloException	{
        C1 = new IloRange[p.getDemand().length];

        //∀i∈I
        for (int i = 0; i < p.getDemand().length; i++) {
            IloLinearNumExpr left = cplex.linearNumExpr();

            Request od = in.getRequestSet().get(i);
            //φ
            for (int k = 0; k < od.getNumberOfLadenPath(); k++) {
                left.addTerm(1, xVar.get(i)[k]);
                left.addTerm(1, yVar.get(i)[k]);
            }

            left.addTerm(1, gVar[i]);

            String ConstrName = "C1(" + (i + 1) + ")";
            C1[i] = cplex.addEq(left, p.getDemand()[i] + p.getMaximumDemandVariation()[i] * uValue[i], ConstrName);
        }
    }

    // Vessel Capacity Constraint :
    // C-6------β
    // C-6<= 0
    private void setConstraint2_1() throws IloException	{
        C2_1 = new IloRange[p.getTravelingArcsSet().length];

        // ∀<n,n'>∈A'
        for(int nn = 0; nn<p.getTravelingArcsSet().length; nn++)
        {
            IloLinearNumExpr left=cplex.linearNumExpr();

            // i∈I
            for(int i=0;i<p.getDemand().length;i++)
            {
                Request od = in.getRequestSet().get(i);

                // φ
                for (int k = 0; k < od.getNumberOfLadenPath(); k++) {
                    int j = od.getLadenPathIndexes()[k];

                    left.addTerm(p.getArcAndPath()[nn][j], xVar.get(i)[k]);
                    left.addTerm(p.getArcAndPath()[nn][j], yVar.get(i)[k]);
                }
            }

            double capacity1 = 0;
            double capacity2 = 0;
            // w∈Ω
            for(int w=0; w<p.getVesselPathSet().length; w++)
            {
                int r = in.getVesselPathSet().get(w).getRouteID() - 1;
                // r(w) = r
                // h \in Hr
                for(int h=0;h<p.getVesselSet().length;h++)
                {
                    // vValue[v][r] : come from solution of master problem
                    capacity1 += p.getArcAndVesselPath()[nn][w]
                            *p.getShipRouteAndVesselPath()[r][w]
                            *p.getVesselTypeAndShipRoute()[h][r]
                            *p.getVesselCapacity()[h]
                            * vVarValue1[h][r];
                }
            }
            String ConstrName = "C6-"+"_"+(nn+1);
            C2_1[nn] = cplex.addLe(left, capacity1, ConstrName);
        }
    }
    private void setConstraint2_2() throws IloException	{
        C2_2 = new IloRange[p.getTravelingArcsSet().length];

        // ∀<n,n'>∈A'
        for(int nn = 0; nn<p.getTravelingArcsSet().length; nn++)
        {
            IloLinearNumExpr left=cplex.linearNumExpr();

            // i∈I
            for(int i=0;i<p.getDemand().length;i++)
            {
                Request od = in.getRequestSet().get(i);

                //θ
                for (int k = 0; k < od.getNumberOfEmptyPath(); k++) {
                    int j = od.getEmptyPathIndexes()[k];

                    left.addTerm(p.getArcAndPath()[nn][j], zVar.get(i)[k]);
                }
            }

            double capacity2 = 0;
            // w∈Ω
            for(int w=0; w<p.getVesselPathSet().length; w++)
            {
                int r = in.getVesselPathSet().get(w).getRouteID() - 1;
                // r(w) = r
                // h \in Hr
                for(int h=0;h<p.getVesselSet().length;h++)
                {
                    // vValue[v][r] : come from solution of master problem
                    capacity2 += p.getArcAndVesselPath()[nn][w]
                            *p.getShipRouteAndVesselPath()[r][w]
                            *p.getVesselTypeAndShipRoute()[h][r]
                            *p.getVesselCapacity()[h]
                            * vVarValue2[h][w];
                }
            }
            String ConstrName = "C6-"+"_"+(nn+1);
            C2_2[nn] = cplex.addLe(left, capacity2, ConstrName);
        }
    }


    // Container Flow Conservation Constraint :
    // calculate the number of available empty self-owned containers at port p at time t
    // or
    // L[p][t] > 0 means all empty self-owned containers that repositioning to other ports (sumZ(out))
    // plus all laden self-owned containers that transport to other ports (sumX(out))
    // should no more than (L[p][t-1] + sumX(in) + sumZ(in))
    // C7------γ
    // C7>=0  =>  -C7 <= 0
    // ( item4 - item1 ) * Z + (item3 - item2) * X <= Lp0
    // Output-X + OutputZ - Input-X - Input-Z <= lp0
    private void setConstraint3() throws IloException	{
        C3 = new IloRange[p.getPortSet().length][p.getTimePointSet().length];

        // ∀p∈P
        for(int pp=0;pp<p.getPortSet().length;pp++)
        {
            // ∀t∈T
            for(int t = 1; t<p.getTimePointSet().length; t++)
            {
                IloLinearNumExpr left=cplex.linearNumExpr();

                // i∈I
                for(int i=0;i<p.getDemand().length;i++)
                {
                    Request od = in.getRequestSet().get(i);

                    // Input Z flow:
                    // (item1)
                    // o(i) == p
                    if(p.getPortSet()[pp].equals(p.getOriginOfDemand()[i]))
                    {
                        //θi
                        for (int k = 0; k < od.getNumberOfEmptyPath(); k++)
                        {
                            int j = od.getEmptyPathIndexes()[k];

                            // <n,n'> ∈A'
                            for (int nn = 0; nn < p.getTravelingArcsSet().length; nn++) {
                                // p(n') == p
                                // 1<= t(n')<= t
                                if (in.getTravelingArcSet().get(nn).getDestinationPort().equals(p.getPortSet()[pp])
                                        && in.getTravelingArcSet().get(nn).getDestinationTime() <= t
                                        && in.getTravelingArcSet().get(nn).getDestinationTime() >= 1)
                                {
                                    left.addTerm(p.getArcAndPath()[nn][j], zVar.get(i)[k]);
                                }
                            }
                        }
                    }

                    // Input flow X
                    // item2
                    // d(i) == p
                    if(p.getPortSet()[pp].equals(p.getDestinationOfDemand()[i]))
                    {
                        for(int k=0;k<od.getNumberOfLadenPath();k++)
                        {
                            int j = od.getLadenPathIndexes()[k];

                            // <n,n'>∈A'
                            for(int nn = 0; nn<p.getTravelingArcsSet().length; nn++)
                            {
                                // p(n‘)∈p
                                // 1 <= t(n')<= t-sp
                                if(in.getTravelingArcSet().get(nn).getDestinationPort().equals(p.getPortSet()[pp])
                                        &&in.getTravelingArcSet().get(nn).getDestinationTime()<=t-p.getTurnOverTime()[pp]
                                        &&in.getTravelingArcSet().get(nn).getDestinationTime()>=1)
                                {
                                    left.addTerm(p.getArcAndPath()[nn][j], xVar.get(i)[k]);
                                }
                            }
                        }
                    }


                    //Output  flow X
                    // item3
                    // o(i) == p
                    if(p.getPortSet()[pp].equals(p.getOriginOfDemand()[i]))
                    {
                        // φi
                        for(int k=0;k<od.getNumberOfLadenPath();k++)
                        {
                            int j = od.getLadenPathIndexes()[k];

                            // <n.n'>∈A'
                            for(int nn = 0; nn<p.getTravelingArcsSet().length; nn++)
                            {
                                //p(n) == p
                                // t(n) <= t
                                if(in.getTravelingArcSet().get(nn).getOriginPort().equals(p.getPortSet()[pp])
                                        &&in.getTravelingArcSet().get(nn).getOriginTime() <=t
                                        &&in.getTravelingArcSet().get(nn).getOriginTime() >=1)
                                {
                                    left.addTerm(-p.getArcAndPath()[nn][j], xVar.get(i)[k]);
                                }
                            }
                        }
                    }


                    // Output Flow Z
                    // item4
                    // θ
                    for(int k=0;k<od.getNumberOfEmptyPath();k++)
                    {
                        int j = od.getEmptyPathIndexes()[k];

                        // <n,n'>∈A'
                        for(int nn = 0; nn<p.getTravelingArcsSet().length; nn++)
                        {
                            // p(n) == p
                            // t(n) <= t
                            if(in.getTravelingArcSet().get(nn).getOriginPort().equals(p.getPortSet()[pp])
                                    && in.getTravelingArcSet().get(nn).getOriginTime()<=t
                                    && in.getTravelingArcSet().get(nn).getOriginTime()>=1)
                            {
                                left.addTerm(-p.getArcAndPath()[nn][j], zVar.get(i)[k]);
                            }
                        }
                    }
                }

                String ConstrName = "C7_"+(pp+1)+"_"+(t);
                C3[pp][t] = cplex.addGe(left, -p.getInitialEmptyContainer()[pp], ConstrName);
            }
        }
    }

    public void solveModel()	{
        try
        {
            if (WhetherExportModel)
                exportModel();
            long startTime = System.currentTimeMillis();
            if (cplex.solve())
            {
                long endTime = System.currentTimeMillis();
                setObjVal(cplex.getObjValue());
                setSolveTime(endTime - startTime);
                setObjGap(cplex.getMIPRelativeGap());
            }
            else
            {
                System.out.println("SubProblem No solution");
            }

        }
        catch (IloException ex) {
            System.out.println("Concert Error: " + ex);
        }
    }

    // change the coefficients of capacity constraints
    public void changeConstraintCoefficients(int[][] vVarValue1, int[][] vVarValue2, int[] uValue) throws IloException {
        this.vVarValue1 = vVarValue1;
        this.vVarValue2 = vVarValue2;
        this.uValue = IntArrayWrapper.IntArrayToDoubleArray(uValue);

        for(int i=0;i<p.getDemand().length;i++)
        {
            C1[i].setBounds(p.getDemand()[i]+p.getMaximumDemandVariation()[i]*uValue[i],
                    p.getDemand()[i]+p.getMaximumDemandVariation()[i]*uValue[i]);
        }

        // ∀<n,n'>∈A'
        for(int nn = 0; nn<p.getTravelingArcsSet().length; nn++)
        {
            double capacity1 = 0;
            double capacity2 = 0;
            // w∈Ω
            for(int w=0; w<p.getVesselPathSet().length; w++)
            {
                int r = in.getVesselPathSet().get(w).getRouteID() - 1;
                // r(w) = r
                // h \in Hr
                for(int h=0;h<p.getVesselSet().length;h++)
                {
                    // vValue[v][r] : come from solution of master problem
                    capacity1 += p.getArcAndVesselPath()[nn][w]
                            *p.getShipRouteAndVesselPath()[r][w]
                            *p.getVesselTypeAndShipRoute()[h][r]
                            *p.getVesselCapacity()[h]
                            * vVarValue1[h][r];
                    // vValue[v][r] : come from solution of master problem
                    capacity2 += p.getArcAndVesselPath()[nn][w]
                            *p.getShipRouteAndVesselPath()[r][w]
                            *p.getVesselTypeAndShipRoute()[h][r]
                            *p.getVesselCapacity()[h]
                            * vVarValue2[h][w];
                }
            }
            C2_1[nn].setBounds(0, capacity1);
            C2_2[nn].setBounds(0, capacity2);
        }
    }
    public void changeConstraintCoefficients(int[][] vVarValue1, int[][] vVarValue2, double[] uValue) throws IloException {
        this.vVarValue1 = vVarValue1;
        this.vVarValue2 = vVarValue2;
        this.uu = uValue;

        for(int i=0;i<p.getDemand().length;i++)        {
            C1[i].setBounds(p.getDemand()[i]+p.getMaximumDemandVariation()[i]*uValue[i],
                    p.getDemand()[i]+p.getMaximumDemandVariation()[i]*uValue[i]);
        }

        // ∀<n,n'>∈A'
        for(int nn = 0; nn<p.getTravelingArcsSet().length; nn++)
        {
            double capacity1 = 0;
            double capacity2 = 0;
            // w∈Ω
            for(int w=0; w<p.getVesselPathSet().length; w++)
            {
                int r = in.getVesselPathSet().get(w).getRouteID() - 1;
                // r(w) = r
                // h \in Hr
                for(int h=0;h<p.getVesselSet().length;h++)
                {
                    // vValue[v][r] : come from solution of master problem
                    capacity1 += p.getArcAndVesselPath()[nn][w]
                            *p.getShipRouteAndVesselPath()[r][w]
                            *p.getVesselTypeAndShipRoute()[h][r]
                            *p.getVesselCapacity()[h]
                            * vVarValue1[h][r];
                    // vValue[v][r] : come from solution of master problem
                    capacity2 += p.getArcAndVesselPath()[nn][w]
                            *p.getShipRouteAndVesselPath()[r][w]
                            *p.getVesselTypeAndShipRoute()[h][r]
                            *p.getVesselCapacity()[h]
                            * vVarValue2[h][w];
                }
            }
            C2_1[nn].setBounds(0, capacity1);
            C2_2[nn].setBounds(0, capacity2);
        }
    }

    public double getLadenCost() {
        return ladenCost;
    }

    public void setLadenCost(double ladenCost) {
        this.ladenCost = ladenCost;
    }

    public double getEmptyCost() {
        return emptyCost;
    }

    public void setEmptyCost(double emptyCost) {
        this.emptyCost = emptyCost;
    }

    public double getPenaltyCost() {
        return penaltyCost;
    }

    public void setPenaltyCost(double penaltyCost) {
        this.penaltyCost = penaltyCost;
    }

    public double getRentalCost() {
        return rentalCost;
    }

    public void setRentalCost(double rentalCost) {
        this.rentalCost = rentalCost;
    }

}

File: Multi_routes_deploy/sources/model/SubProblemReactive.cpp
#include "SubProblemReactive.h"

namespace fleetdeployment {

    SubProblemReactive::SubProblemReactive(const InputData& in, const Parameter& p) 
        : in_(in), p_(p)
    {
        try {
            env_ = IloEnv();
            model_ = IloModel(env_);
            cplex_ = IloCplex(model_);

            cplex_.setOut(env_.getNullStream());
            cplex_.setParam(IloCplex::Param::WorkMem, MaxWorkMem);
            cplex_.setParam(IloCplex::Param::TimeLimit, MIPTimeLimit);
            cplex_.setParam(IloCplex::Param::MIP::Tolerances::MIPGap, MIPGapLimit);
            cplex_.setParam(IloCplex::Param::Threads, MaxThreads);

            SetDecisionVars();
            SetObjective();
            SetConstraints();
        }
        catch (const IloException& e) {
            std::cerr << "Error: " << e.getMessage() << std::endl;
        }

        // 使用 std::vector 初始化 vVarValue1, vVarValue2 和 uValue
        vVarValue1_.resize(p_.GetVesselSet().size(), std::vector<int>(p_.GetVesselRouteSet().size()));
        vVarValue2_.resize(p_.GetVesselSet().size(), std::vector<int>(p_.GetVesselPathSet().size()));
        uValue_.resize(p_.GetDemand().size());
    }



    void SubProblemReactive::SetDecisionVars() {
        // Assuming Request, InputData, Parameter classes are defined appropriately
        // Resize vectors according to the number of demands
        xVar_.resize(p_.GetDemand().size());
        yVar_.resize(p_.GetDemand().size());
        zVar_.resize(p_.GetDemand().size());
        gVar_.resize(p_.GetDemand().size());

        std::string varName;

        for (int i = 0; i < p_.GetDemand().size(); ++i) {
            Request od = in_.GetRequests()[i];

            xVar_[i].resize(od.GetNumberOfLadenPath());
            yVar_[i].resize(od.GetNumberOfLadenPath());
            zVar_[i].resize(od.GetNumberOfEmptyPath());

            for (int k = 0; k < od.GetNumberOfLadenPath(); ++k) {
                varName = "x(" + std::to_string(i + 1) + ")";
                xVar_[i][k] = IloNumVar(env_, 0, IloInfinity, varName.c_str());
                varName = "y(" + std::to_string(i + 1) + ")";
                yVar_[i][k] = IloNumVar(env_, 0, IloInfinity, varName.c_str());
            }
            for (int k = 0; k < od.GetNumberOfEmptyPath(); ++k) {
                varName = "z(" + std::to_string(i + 1) + ")";
                zVar_[i][k] = IloNumVar(env_, 0, IloInfinity, varName.c_str());
            }

            varName = "g(" + std::to_string(i + 1) + ")";
            gVar_[i] = IloNumVar(env_, 0, IloInfinity, varName.c_str());
        }
    }


    // Minimize total cost about containers
    void SubProblemReactive::SetObjective() {
        IloExpr obj(env_);

        // Iterate over demands
        for (int i = 0; i < p_.GetDemand().size(); ++i) {
            // Penalty cost for unsatisfied demand
            obj += p_.GetPenaltyCostForDemand()[i] * gVar_[i];

            Request od = in_.GetRequests()[i];
            // Iterate over laden paths
            for (int k = 0; k < od.GetNumberOfLadenPath(); ++k) {
                int j = od.GetLadenPathIndexes()[k];
                // Demurrage of self-owned and leased containers, and rental cost on laden paths
                obj += p_.GetLadenPathCost()[j] * xVar_[i][k];
                obj += p_.GetLadenPathCost()[j] * yVar_[i][k];
                obj += p_.GetRentalCost() * p_.GetTravelTimeOnPath()[j] * yVar_[i][k];
            }

            // Iterate over empty paths
            for (int k = 0; k < od.GetNumberOfEmptyPath(); ++k) {
                int j = od.GetEmptyPathIndexes()[k];
                // Demurrage of self-owned containers for empty path
                obj += p_.GetEmptyPathCost()[j] * zVar_[i][k];
            }
        }

        model_.add(IloMinimize(env_, obj));
    }


    void SubProblemReactive::SetConstraints() {
        SetConstraint1();
        SetConstraint2_1();
        SetConstraint2_2();
        SetConstraint3();
    }

    // Demand equation :
    // C-5------α
    // C-5 = 0
#include <ilcplex/ilocplex.h>
#include <string>

    void SubProblemReactive::SetConstraint1() {
        // Assuming C1_ is a std::vector<IloRange> member of SubProblemReactive
        C1_.resize(p_.GetDemand().size());

        // ∀i∈I
        for (int i = 0; i < p_.GetDemand().size(); ++i) {
            IloEnv env = cplex_.getEnv();
            IloExpr left(env_);

            Request od = in_.GetRequests()[i];
            // φ
            for (int k = 0; k < od.GetNumberOfLadenPath(); ++k) {
                left += xVar_[i][k];
                left += yVar_[i][k];
            }

            left += gVar_[i];

            std::string constr_name = "C1(" + std::to_string(i + 1) + ")";
            C1_[i] = IloRange(env_, p_.GetDemand()[i] + p_.GetMaximumDemandVariation()[i] * uValue_[i], left, p_.GetDemand()[i] + p_.GetMaximumDemandVariation()[i] * uValue_[i], constr_name.c_str());
            // Add the constraint to the model
            cplex_.getModel().add(C1_[i]);
        }
    }


    // Vessel Capacity Constraint :
    // C-6------β
    // C-6<= 0
    void SubProblemReactive::SetConstraint2_1() {
        C2_1_.resize(p_.GetTravelArcsSet().size());

        for (int nn = 0; nn < p_.GetTravelArcsSet().size(); ++nn) {
            IloEnv env = cplex_.getEnv();
            IloExpr left(env_);

            for (int i = 0; i < p_.GetDemand().size(); ++i) {
                Request od = in_.GetRequests()[i];
                for (int k = 0; k < od.GetNumberOfLadenPath(); ++k) {
                    int j = od.GetLadenPathIndexes()[k];
                    left += p_.GetArcAndPath()[nn][j] * xVar_[i][k];
                    left += p_.GetArcAndPath()[nn][j] * yVar_[i][k];
                }
            }

            double capacity1 = 0;
            for (int w = 0; w < p_.GetVesselPathSet().size(); ++w) {
                int r = in_.GetVesselPathSet()[w].GetRouteID() - 1;
                for (int h = 0; h < p_.GetVesselSet().size(); ++h) {
                    capacity1 += p_.GetArcAndVesselPath()[nn][w]
                        * p_.GetShipRouteAndVesselPath()[r][w]
                        * p_.GetVesselTypeAndShipRoute()[h][r]
                        * p_.GetVesselCapacity()[h]
                        * vVarValue1_[h][r];
                }
            }

            std::string constr_name = "C6-_" + std::to_string(nn + 1);
            C2_1_[nn] = IloRange(env, left, capacity1, constr_name.c_str());
            cplex_.getModel().add(C2_1_[nn]);
        }
    }

    void SubProblemReactive::SetConstraint2_2() {
        C2_2_.resize(p_.GetTravelArcsSet().size());

        for (int nn = 0; nn < p_.GetTravelArcsSet().size(); ++nn) {
            IloEnv env = cplex_.getEnv();
            IloExpr left(env_);

            for (int i = 0; i < p_.GetDemand().size(); ++i) {
                Request od = in_.GetRequests()[i];
                for (int k = 0; k < od.GetNumberOfEmptyPath(); ++k) {
                    int j = od.GetEmptyPathIndexes()[k];
                    left += p_.GetArcAndPath()[nn][j] * zVar_[i][k];
                }
            }

            double capacity2 = 0;
            for (int w = 0; w < p_.GetVesselPathSet().size(); ++w) {
                int r = in_.GetVesselPathSet()[w].GetRouteID() - 1;
                for (int h = 0; h < p_.GetVesselSet().size(); ++h) {
                    capacity2 += p_.GetArcAndVesselPath()[nn][w]
                        * p_.GetShipRouteAndVesselPath()[r][w]
                        * p_.GetVesselTypeAndShipRoute()[h][r]
                        * p_.GetVesselCapacity()[h]
                        * vVarValue2_[h][w];
                }
            }

            std::string constr_name = "C6-_" + std::to_string(nn + 1);
            C2_2_[nn] = IloRange(env, left, capacity2, constr_name.c_str());
            cplex_.getModel().add(C2_2_[nn]);
        }
    }



    // Container Flow Conservation Constraint :
    // calculate the number of available empty self-owned containers at port p at time t
    // or
    // L[p][t] > 0 means all empty self-owned containers that repositioning to other ports (sumZ(out))
    // plus all laden self-owned containers that transport to other ports (sumX(out))
    // should no more than (L[p][t-1] + sumX(in) + sumZ(in))
    // C7------γ
    // C7>=0  =>  -C7 <= 0
    // ( item4 - item1 ) * Z + (item3 - item2) * X <= Lp0
    // Output-X + OutputZ - Input-X - Input-Z <= lp0
#include <ilcplex/ilocplex.h>
#include <string>
#include <vector>

    void SubProblemReactive::SetConstraint3() {
        C3_.resize(p_.GetPortSet().size());
        for (auto& row : C3_) {
            row.resize(p_.GetTimePointSet().size());
        }

        for (int pp = 0; pp < p_.GetPortSet().size(); ++pp) {
            for (int t = 1; t < p_.GetTimePointSet().size(); ++t) {
                IloEnv env = cplex_.getEnv();
                IloExpr left(env_);

                for (int i = 0; i < p_.GetDemand().size(); ++i) {
                    Request od = in_.GetRequests()[i];

                    // Input Z flow
                    if (p_.GetPortSet()[pp] == p_.GetOriginOfDemand()[i]) {
                        for (int k = 0; k < od.GetNumberOfEmptyPath(); ++k) {
                            int j = od.GetEmptyPathIndexes()[k];
                            for (int nn = 0; nn < p_.GetTravelArcsSet().size(); ++nn) {
                                if (in_.GetTravelArcs()[nn].GetDestinationPort() == p_.GetPortSet()[pp]
                                    && in_.GetTravelArcs()[nn].GetDestinationTime() <= t
                                    && in_.GetTravelArcs()[nn].GetDestinationTime() >= 1) {
                                    left += p_.GetArcAndPath()[nn][j] * zVar_[i][k];
                                }
                            }
                        }
                    }

                    // Input flow X
                    if (p_.GetPortSet()[pp] == p_.GetDestinationOfDemand()[i]) {
                        for (int k = 0; k < od.GetNumberOfLadenPath(); ++k) {
                            int j = od.GetLadenPathIndexes()[k];
                            for (int nn = 0; nn < p_.GetTravelArcsSet().size(); ++nn) {
                                if (in_.GetTravelArcs()[nn].GetDestinationPort() == p_.GetPortSet()[pp]
                                    && in_.GetTravelArcs()[nn].GetDestinationTime() <= t - p_.GetTurnOverTimeSet()[pp]
                                    && in_.GetTravelArcs()[nn].GetDestinationTime() >= 1) {
                                    left += p_.GetArcAndPath()[nn][j] * xVar_[i][k];
                                }
                            }
                        }
                    }

                    // Output flow X
                    if (p_.GetPortSet()[pp] == p_.GetOriginOfDemand()[i]) {
                        for (int k = 0; k < od.GetNumberOfLadenPath(); ++k) {
                            int j = od.GetLadenPathIndexes()[k];
                            for (int nn = 0; nn < p_.GetTravelArcsSet().size(); ++nn) {
                                if (in_.GetTravelArcs()[nn].GetOriginPort() == p_.GetPortSet()[pp]
                                    && in_.GetTravelArcs()[nn].GetOriginTime() <= t
                                    && in_.GetTravelArcs()[nn].GetOriginTime() >= 1) {
                                    left -= p_.GetArcAndPath()[nn][j] * xVar_[i][k];
                                }
                            }
                        }
                    }

                    // Output Flow Z
                    for (int k = 0; k < od.GetNumberOfEmptyPath(); ++k) {
                        int j = od.GetEmptyPathIndexes()[k];
                        for (int nn = 0; nn < p_.GetTravelArcsSet().size(); ++nn) {
                            if (in_.GetTravelArcs()[nn].GetOriginPort() == p_.GetPortSet()[pp]
                                && in_.GetTravelArcs()[nn].GetOriginTime() <= t
                                && in_.GetTravelArcs()[nn].GetOriginTime() >= 1) {
                                left -= p_.GetArcAndPath()[nn][j] * zVar_[i][k];
                            }
                        }
                    }
                }

                std::string constr_name = "C7_" + std::to_string(pp + 1) + "_" + std::to_string(t);
                C3_[pp][t] = IloRange(env, -p_.GetInitialEmptyContainer()[pp], left, IloInfinity, constr_name.c_str());
                cplex_.getModel().add(C3_[pp][t]);
            }
        }
    }


#include <ilcplex/ilocplex.h>
#include <iostream>

    void SubProblemReactive::solveModel() {
        try {
            if (cplex_.solve()) {
                SetObjVal(cplex_.getObjValue());
            }
            else {
                std::cout << "SubProblem No solution" << std::endl;
            }
        }
        catch (const IloException& ex) {
            std::cout << "Concert Error: " << ex.getMessage() << std::endl;
        }
    }



    // change the coefficients of capacity constraints
#include <vector>

    void SubProblemReactive::changeConstraintCoefficients(const std::vector<std::vector<int>>& vVarValue1, const std::vector<std::vector<int>>& vVarValue2, const std::vector<double>& uValue) {
        vVarValue1_ = vVarValue1;
        vVarValue2_ = vVarValue2;
        uValue_ = uValue;

        for (int i = 0; i < p_.GetDemand().size(); ++i) {
            double newBound = p_.GetDemand()[i] + p_.GetMaximumDemandVariation()[i] * uValue_[i];
            C1_[i].setBounds(newBound, newBound);
        }

        for (int nn = 0; nn < p_.GetTravelArcsSet().size(); ++nn) {
            double capacity1 = 0;
            double capacity2 = 0;
            for (int w = 0; w < p_.GetVesselPathSet().size(); ++w) {
                int r = in_.GetVesselPathSet()[w].GetRouteID() - 1;
                for (int h = 0; h < p_.GetVesselSet().size(); ++h) {
                    capacity1 += p_.GetArcAndVesselPath()[nn][w] * p_.GetShipRouteAndVesselPath()[r][w] * p_.GetVesselTypeAndShipRoute()[h][r] * p_.GetVesselCapacity()[h] * vVarValue1_[h][r];
                    capacity2 += p_.GetArcAndVesselPath()[nn][w] * p_.GetShipRouteAndVesselPath()[r][w] * p_.GetVesselTypeAndShipRoute()[h][r] * p_.GetVesselCapacity()[h] * vVarValue2_[h][w];
                }
            }
            C2_1_[nn].setBounds(0, capacity1);
            C2_2_[nn].setBounds(0, capacity2);
        }
    }


}